### ISSUE TITLE
[question] `self.dependencies[...].package_folder` in package_info()

### ISSUE THREAD

**Issue by sykhro** [conan-io/conan#14829]
### What is your question?

[The docs](https://docs.conan.io/2/reference/conanfile/methods/generate.html#dependencies-interface) wanted me to open an issue :) 

I'm packaging clang for (cross-)compiling for Windows. I have always been using this package as a `tool_requires` inside the host profile.

Now, to make this work universally, I have added a sysroot package for Visual Studio which contains the MSVCRT, WinSDK, DIA SDK. Outside of conan, this setup is well-tested and working for cross-compiling from Linux or compiling from windows without a specific VS version installed.

I'm handling this as follows:
```
def requirements(self):
        if self.settings.os == "Windows":
            self.requires(self._VISUALSTUDIO_SYSROOT_REFERENCE, build=False)
```
this makes sense since the host profile is Windows and these are all libraries needed by that.

Here comes the problem:
```python
def package_info():
          # This is used for setting the profile info automatically by the package
          [...]
            vs_sysroot = Path(self.dependencies[self._VISUALSTUDIO_SYSROOT_REFERENCE].package_folder)
            cflags = [
                "/permissive-",  # No microsoftisms
                "/Zc:preprocessor",  # Use the standard conforming preprocessor
                "/volatile:iso",  # Use the standard conforming volatile behaviour
                "-fuse-ld=lld",  # this is implicitly lld-link if the frontend is clang-cl
                f'/winsdkdir {vs_sysroot.joinpath("sdk").resolve().as_posix()}'
                f'/vctoolsdir {vs_sysroot.joinpath("crt").resolve().as_posix()}',
            ]
            self.conf_info.define("tools.build:cflags", cflags)
            self.conf_info.define("tools.build:cxxflags", cflags)
```
the first line breaks[1] since `package_folder` is `None` _when this clang package is used as tool_requires_.

If I understand the issue correctly, a workaround could be to have these flags be set by the sysroot package itself.
I think there was a similar issue about this already but I can't find it.

1: this is the case when using a test_package that has clang in its tool_requires. I'd need that to ensure the packaged clang works. I'll also try from a profile.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @sykhro 

Thanks for your feedback.
It would be good to know what command output you are getting before the above fails due to ``package_folder=None``. This can happen when a binary is "Skipped" because it is not considered necessary. If the dependency information is necessary at runtime for ``package_info()`` to work, maybe this can be solved by declaring that there is a runtime requirement to the dependency: ``self.requires(self._VISUALSTUDIO_SYSROOT_REFERENCE, run=True)`` (the ``build=False`` is the default, not necessary)

In other words, in order for ``clang`` package to be able run successfully, it needs the ``visual_runtime`` package to be installed and available at ``clang`` runtime. 

Can you please give that a try and see if it works?

**Comment by sykhro** [comment]
I'm sorry! I tried this right away and forgot to get back to you because the build took ages. Your suggestion worked, thanks :)

**Comment by memsharded** [comment]
No prob :) Thanks for following up and the feedback!

**Comment by sykhro** [comment]
I have just noticed that it resolves to the cache folder instead of the deploy folder.
In `package_info()` I have:
```python
vs_sysroot = Path(
    self.dependencies[self._VISUALSTUDIO_SYSROOT_REFERENCE].package_folder
)
self.conf_info.define(
  "tools.build:cxxflags",
  [f'/winsdkdir {vs_sysroot.joinpath("sdk").resolve().as_posix()}',
   f'/vctoolsdir {vs_sysroot.joinpath("crt").resolve().as_posix()}']
)
```
which results in the following `cmaketoolchain.cmake`:
```CMake
string(APPEND CONAN_CXX_FLAGS " [...] /winsdkdir C:/Users/Erin/.conan2/p/visuab996ce459d997/p/sdk /vctoolsdir C:/Users/Erin/.conan2/p/visuab996ce459d997/p/crt [...]")
```

**Comment by memsharded** [comment]
I think the issue could be that you need to explicitly define it as "path" with ``conf_info.define_path()`` in order to make Conan aware that it is a path and it should be handled. Could you please try that?

Update: Sorry, I didn't realize that you are defining `` "tools.build:cxxflags",``. Yes, these will not work, it is not possible for Conan to make these flags process the folders in that way to make the packages relocatable. You would need to somehow abstract that information and compose the ``/winsdkdir`` and ``/vctoolsdir`` in the consumer side.

**Comment by sykhro** [comment]
> I think the issue could be that you need to explicitly define it as "path" with `conf_info.define_path()` in order to make Conan aware that it is a path and it should be handled. Could you please try that?
> 
> Update: Sorry, I didn't realize that you are defining ` "tools.build:cxxflags",`. Yes, these will not work, it is not possible for Conan to make these flags process the folders in that way to make the packages relocatable. You would need to somehow abstract that information and compose the `/winsdkdir` and `/vctoolsdir` in the consumer side.

I figured. I assume doing `conf_info.define_path()` and then `conf_info.get()` right after wouldn't help?
In the worst case I guess I can write a custom deployer for this

**Comment by memsharded** [comment]
Sounds a custom generator would be more adequate maybe, something you can use in the consumers ``generate()``

though I still think it would be nice to have something that works without the consumers having to add something explicitly in the consumer recipes in order to use this as ``tool_requires``. Are you using always CMake in your consumer recipes? 


> I have just noticed that it resolves to the cache folder instead of the deploy folder.

It seems that you are using the Conan packages in the way described in https://blog.conan.io/2023/05/23/Conan-agnostic-deploy-dependencies.html, I am curious to learn what is your use case to deploy the dependencies and not using them directly in the cache. Are you distributing a Conan-independent SDK with all deps to developers?

**Comment by sykhro** [comment]
> Sounds a custom generator would be more adequate maybe, something you can use in the consumers generate()

Didn't think about that, it does sound cleaner

> Are you using always CMake in your consumer recipes?

Yes, for now, at least

> Are you distributing a Conan-independent SDK with all deps to developers?

Essentially, yes. We are developing a language runtime which happens to have external dependencies. The dependencies are also needed by the compiler in the final binary generation step, having a relocatable archive with every library versioned properly is very useful

**Comment by memsharded** [comment]
> Are you using always CMake in your consumer recipes?

Maybe it is worth trying some cmake-specific approach, for example, creating a ``user_toolchain`` in the tool-require package (inspired by https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/inject_cmake_variables.html, but not the same), that contains the definition of flags, and makes the paths relative to the current "myusertoolchain.cmake", because that will be inside the package. If the file is deployed together with the binaries, it will keep pointing to the deployed binaries because the path is relative. And the ``CMakeToolchain`` should be able to relativize the path to the package-folder containing the user-toolchain.

**Comment by AbrilRBS** [comment]
Closing as solved with the comments above. Feel free to create new issues if anyone finds themselves in similar situations, thanks!
